File: nest-ts-bun-backend/tests/request-context.test.ts
import type { INestApplication } from '@nestjs/common';
import request from 'supertest';
import { afterEach, describe, expect, it } from 'vitest';
import { createApp } from '../src/app';

describe('request context', () => {
  let app: INestApplication | undefined;

  afterEach(async () => {
    if (app) {
      await app.close();
      app = undefined;
    }
  });

  it('propagates incoming x-request-id and emits trace headers', async () => {
    app = await createApp();
    const requestId = 'req-12345';

    const response = await request(app.getHttpServer())
      .get('/api/v1/ping')
      .set('x-request-id', requestId);

    expect(response.status).toBe(200);
    expect(response.headers['x-request-id']).toBe(requestId);
    expect(response.headers['x-trace-id']).toBeDefined();
    expect(response.headers['x-span-id']).toBeDefined();
  });

  it('adds request and trace identifiers to error responses', async () => {
    app = await createApp();

    const response = await request(app.getHttpServer()).get('/route-does-not-exist');

    expect(response.status).toBe(404);
    expect(response.body.requestId).toBeDefined();
    expect(response.body.traceId).toBeDefined();
  });
});

File: flask-python-backend/tests/request_context_test.py
def test_propagates_request_id_and_trace_headers(client) -> None:
    request_id = 'req-12345'

    response = client.get('/api/v1/ping', headers={'x-request-id': request_id})

    assert response.status_code == 200
    assert response.headers['x-request-id'] == request_id
    assert response.headers['x-trace-id']
    assert response.headers['x-span-id']


def test_trace_identifiers_appear_in_error_response(client) -> None:
    response = client.get('/route-does-not-exist')
    payload = response.get_json()

    assert response.status_code == 404
    assert payload['requestId']
    assert payload['traceId']
    assert payload['spanId']
